Could not get unknown property 'idea' for root project

Hi,

I cannot make this plugin work on my project, i always get the same error during build : 
` Could not get unknown property 'idea' for root project`

Any thoughts ?

Hi and thanks to take care of my problem, i'm just running a build after adding plugin in my build.gradle : 

`./gradlew clean build --stacktrace`

`org.gradle.api.plugins.InvalidPluginException: An exception occurred applying plugin request [id: 'org.inferred.processors', version: '1.2.5']
        at org.gradle.plugin.use.internal.DefaultPluginRequestApplicator.applyPlugin(DefaultPluginRequestApplicator.java:224)
        at org.gradle.plugin.use.internal.DefaultPluginRequestApplicator.applyPlugins(DefaultPluginRequestApplicator.java:171)
        at org.gradle.configuration.DefaultScriptPluginFactory$ScriptPluginImpl.apply(DefaultScriptPluginFactory.java:156)
        at org.gradle.configuration.project.BuildScriptProcessor.execute(BuildScriptProcessor.java:39)
        at org.gradle.configuration.project.BuildScriptProcessor.execute(BuildScriptProcessor.java:26)
        at org.gradle.configuration.project.ConfigureActionsProjectEvaluator.evaluate(ConfigureActionsProjectEvaluator.java:34)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.doConfigure(LifecycleProjectEvaluator.java:70)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.access$000(LifecycleProjectEvaluator.java:33)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$1.execute(LifecycleProjectEvaluator.java:53)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$1.execute(LifecycleProjectEvaluator.java:50)
        at org.gradle.internal.Transformers$4.transform(Transformers.java:169)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:106)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:61)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.evaluate(LifecycleProjectEvaluator.java:50)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:628)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:129)
        at org.gradle.execution.TaskPathProjectEvaluator.configure(TaskPathProjectEvaluator.java:35)
        at org.gradle.execution.TaskPathProjectEvaluator.configureHierarchy(TaskPathProjectEvaluator.java:60)
        at org.gradle.configuration.DefaultBuildConfigurer.configure(DefaultBuildConfigurer.java:38)
        at org.gradle.initialization.DefaultGradleLauncher$1.execute(DefaultGradleLauncher.java:161)
        at org.gradle.initialization.DefaultGradleLauncher$1.execute(DefaultGradleLauncher.java:158)
        at org.gradle.internal.Transformers$4.transform(Transformers.java:169)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:106)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:56)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:158)
        at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:119)
        at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:102)
        at org.gradle.launcher.exec.GradleBuildController.run(GradleBuildController.java:71)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:41)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:26)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:75)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:49)
        at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:44)
        at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:29)
        at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:67)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:47)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:26)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:34)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:74)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:72)
        at org.gradle.util.Swapper.swap(Swapper.java:38)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:72)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:60)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:72)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:50)
        at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:297)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
Caused by: org.gradle.api.internal.plugins.PluginApplicationException: Failed to apply plugin [id 'org.inferred.processors']
        at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:155)
        at org.gradle.api.internal.plugins.DefaultPluginManager.apply(DefaultPluginManager.java:112)
        at org.gradle.plugin.use.internal.DefaultPluginRequestApplicator$5.run(DefaultPluginRequestApplicator.java:174)
        at org.gradle.plugin.use.internal.DefaultPluginRequestApplicator.applyPlugin(DefaultPluginRequestApplicator.java:212)
        ... 61 more
Caused by: groovy.lang.MissingPropertyException: Could not get unknown property 'idea' for root project 'XXX' of type org.gradle.api.Project.
        at org.gradle.internal.metaobject.AbstractDynamicObject.getMissingProperty(AbstractDynamicObject.java:85)
        at org.gradle.internal.metaobject.AbstractDynamicObject.getProperty(AbstractDynamicObject.java:62)
        at org.gradle.api.internal.project.DefaultProject_Decorated.getProperty(Unknown Source)
        at org.inferred.gradle.ProcessorsPlugin$_updateIdeaCompilerConfiguration_closure11$_closure35$_closure36.doCall(ProcessorsPlugin.groovy:209)
        at org.inferred.gradle.ProcessorsPlugin$_updateIdeaCompilerConfiguration_closure11$_closure35$_closure36.doCall(ProcessorsPlugin.groovy)
        at org.inferred.gradle.ProcessorsPlugin$_updateIdeaCompilerConfiguration_closure11$_closure35.doCall(ProcessorsPlugin.groovy:208)
        at org.inferred.gradle.ProcessorsPlugin$_updateIdeaCompilerConfiguration_closure11$_closure35.doCall(ProcessorsPlugin.groovy)
        at org.inferred.gradle.ProcessorsPlugin$_updateIdeaCompilerConfiguration_closure11.doCall(ProcessorsPlugin.groovy:207)
        at org.inferred.gradle.ProcessorsPlugin$_updateIdeaCompilerConfiguration_closure11.doCall(ProcessorsPlugin.groovy)
        at org.inferred.gradle.ProcessorsPlugin.updateIdeaCompilerConfiguration(ProcessorsPlugin.groovy:206)
        at org.inferred.gradle.ProcessorsPlugin$updateIdeaCompilerConfiguration.callStatic(Unknown Source)
        at org.inferred.gradle.ProcessorsPlugin.apply(ProcessorsPlugin.groovy:118)
        at org.inferred.gradle.ProcessorsPlugin.apply(ProcessorsPlugin.groovy)
        at org.gradle.api.internal.plugins.ImperativeOnlyPluginApplicator.applyImperative(ImperativeOnlyPluginApplicator.java:35)
        at org.gradle.api.internal.plugins.RuleBasedPluginApplicator.applyImperative(RuleBasedPluginApplicator.java:43)
        at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:139)
        ... 64 more

`

--- a/build.gradle
+++ b/build.gradle
@@ -39,6 +39,12 @@ dependencies {
   testRuntime files(createClasspathManifest)
 }
 
+test {
+  testLogging {
+    exceptionFormat = 'full'
+  }
+}
+
 //// Publication //////////////////////////////////////////////////////////////////////
 
 group = 'org.inferred'

--- a/src/main/groovy/org/inferred/gradle/ProcessorsPlugin.groovy
+++ b/src/main/groovy/org/inferred/gradle/ProcessorsPlugin.groovy
@@ -111,15 +111,17 @@ class ProcessorsPlugin implements Plugin<Project> {
       })
     })
 
-    // If the project uses .idea directory structure, update compiler.xml directly
-    File ideaCompilerXml = project.file('.idea/compiler.xml')
-    if (ideaCompilerXml.isFile()) {
-      Node parsedProjectXml = (new XmlParser()).parse(ideaCompilerXml)
-      updateIdeaCompilerConfiguration(project, parsedProjectXml)
-      ideaCompilerXml.withWriter { writer ->
-        XmlNodePrinter nodePrinter = new XmlNodePrinter(new PrintWriter(writer))
-        nodePrinter.setPreserveWhitespace(true)
-        nodePrinter.print(parsedProjectXml)
+    project.afterEvaluate {
+      // If the project uses .idea directory structure, update compiler.xml directly
+      File ideaCompilerXml = project.file('.idea/compiler.xml')
+      if (ideaCompilerXml.isFile()) {
+        Node parsedProjectXml = (new XmlParser()).parse(ideaCompilerXml)
+        updateIdeaCompilerConfiguration(project, parsedProjectXml)
+        ideaCompilerXml.withWriter { writer ->
+          XmlNodePrinter nodePrinter = new XmlNodePrinter(new PrintWriter(writer))
+          nodePrinter.setPreserveWhitespace(true)
+          nodePrinter.print(parsedProjectXml)
+        }
       }
     }
 
@@ -203,11 +205,18 @@ class ProcessorsPlugin implements Plugin<Project> {
       new Node(compilerConfiguration, "annotationProcessing")
     }
 
+    def outputDir = 'generated_src'
+    def testOutputDir = 'generated_testSrc'
+    if (project.hasProperty('idea')) {
+      outputDir = project.idea.processors.outputDir
+      testOutputDir = project.idea.processors.testOutputDir
+    }
+
     compilerConfiguration.annotationProcessing.replaceNode{
       annotationProcessing() {
         profile(default: 'true', name: 'Default', enabled: 'true') {
-          sourceOutputDir(name: project.idea.processors.outputDir)
-          sourceTestOutputDir(name: project.idea.processors.testOutputDir)
+          sourceOutputDir(name: outputDir)
+          sourceTestOutputDir(name: testOutputDir)
           outputRelativeToContentRoot(value: 'true')
           processorPath(useClasspath: 'true')
         }

--- a/src/test/groovy/org/inferred/gradle/ProcessorsPluginFunctionalTest.groovy
+++ b/src/test/groovy/org/inferred/gradle/ProcessorsPluginFunctionalTest.groovy
@@ -595,10 +595,9 @@ public class ProcessorsPluginFunctionalTest {
             .withArguments("idea", "--stacktrace")
             .build()
 
-    def xml = new XmlSlurper().parse(testProjectDirRoot.toPath().resolve(".idea/compiler.xml").toFile())
+    def xml = testProjectDirRoot.toPath().resolve(".idea/compiler.xml").toFile().text.trim()
 
-    def expected = new XmlSlurper().parseText("""
-      <?xml version="1.0" encoding="UTF-8"?>
+    def expected = """
       <project version="4">
         <component name="CompilerConfiguration">
           <annotationProcessing>
@@ -611,7 +610,7 @@ public class ProcessorsPluginFunctionalTest {
           </annotationProcessing>
         </component>
       </project>
-    """.trim())
+    """.stripIndent().trim()
 
     assertEquals(expected, xml)
   }
@@ -638,10 +637,9 @@ public class ProcessorsPluginFunctionalTest {
             .withArguments("idea", "--stacktrace")
             .build()
 
-    def xml = new XmlSlurper().parse(testProjectDirRoot.toPath().resolve(".idea/compiler.xml").toFile())
+    def xml = testProjectDirRoot.toPath().resolve(".idea/compiler.xml").toFile().text.trim()
 
-    def expected = new XmlSlurper().parseText("""
-      <?xml version="1.0" encoding="UTF-8"?>
+    def expected = """
       <project version="4">
         <component name="CompilerConfiguration">
           <annotationProcessing>
@@ -654,7 +652,7 @@ public class ProcessorsPluginFunctionalTest {
           </annotationProcessing>
         </component>
       </project>
-    """.trim())
+    """.stripIndent().trim()
 
     assertEquals(expected, xml)
   }

--- a/src/test/groovy/org/inferred/gradle/ProcessorsPluginFunctionalTest.groovy
+++ b/src/test/groovy/org/inferred/gradle/ProcessorsPluginFunctionalTest.groovy
@@ -658,6 +658,54 @@ public class ProcessorsPluginFunctionalTest {
   }
 
   @Test
+  public void testUserSpecifiedDirectoriesUsedInIdeaCompilerXml() throws IOException {
+    buildFile << """
+      apply plugin: 'java'
+      apply plugin: 'idea'
+      apply plugin: 'org.inferred.processors'
+
+      idea.processors {
+        outputDir = 'foo'
+        testOutputDir = 'bar'
+      }
+    """
+
+    new File(testProjectDir.newFolder('.idea'), 'compiler.xml') << """
+      <?xml version="1.0" encoding="UTF-8"?>
+      <project version="4">
+        <component name="CompilerConfiguration">
+          <annotationProcessing/>
+        </component>
+      </project>
+    """.trim()
+
+    File testProjectDirRoot = testProjectDir.getRoot()
+    GradleRunner.create()
+            .withProjectDir(testProjectDirRoot)
+            .withArguments("idea", "--stacktrace")
+            .build()
+
+    def xml = testProjectDirRoot.toPath().resolve(".idea/compiler.xml").toFile().text.trim()
+
+    def expected = """
+      <project version="4">
+        <component name="CompilerConfiguration">
+          <annotationProcessing>
+            <profile default="true" name="Default" enabled="true">
+              <sourceOutputDir name="foo"/>
+              <sourceTestOutputDir name="bar"/>
+              <outputRelativeToContentRoot value="true"/>
+              <processorPath useClasspath="true"/>
+            </profile>
+          </annotationProcessing>
+        </component>
+      </project>
+    """.stripIndent().trim()
+
+    assertEquals(expected, xml)
+  }
+
+  @Test
   public void testOnlyApplyToSubProject() {
     testProjectDir.newFolder("projectA")
     testProjectDir.newFolder("projectB")
@@ -743,6 +791,98 @@ public class ProcessorsPluginFunctionalTest {
         .build()
     assertAutoValueInFile(new File(runner.projectDir, ".classpath"))
     assertAutoValueInFile(new File(runner.projectDir, ".factorypath"))
+  }
+
+  /** @see https://github.com/palantir/gradle-processors/issues/28 */
+  @Test
+  public void testIdeaCompilerConfigurationUpdatedWithoutNeedToApplyIdeaPlugin() throws IOException {
+    buildFile << """
+      apply plugin: 'java'
+      apply plugin: 'org.inferred.processors'
+    """
+
+    new File(testProjectDir.newFolder('.idea'), 'compiler.xml') << """
+      <?xml version="1.0" encoding="UTF-8"?>
+      <project version="4">
+        <component name="CompilerConfiguration">
+          <annotationProcessing/>
+        </component>
+      </project>
+    """.trim()
+
+    File testProjectDirRoot = testProjectDir.getRoot()
+    GradleRunner.create()
+            .withProjectDir(testProjectDirRoot)
+            .withArguments("tasks", "--stacktrace")
+            .build()
+
+    def xml = testProjectDirRoot.toPath().resolve(".idea/compiler.xml").toFile().text.trim()
+
+    def expected = """
+      <project version="4">
+        <component name="CompilerConfiguration">
+          <annotationProcessing>
+            <profile default="true" name="Default" enabled="true">
+              <sourceOutputDir name="generated_src"/>
+              <sourceTestOutputDir name="generated_testSrc"/>
+              <outputRelativeToContentRoot value="true"/>
+              <processorPath useClasspath="true"/>
+            </profile>
+          </annotationProcessing>
+        </component>
+      </project>
+    """.stripIndent().trim()
+
+    assertEquals(expected, xml)
+  }
+
+  /** @see https://github.com/palantir/gradle-processors/issues/53 */
+  @Test
+  public void testCompilerXmlModificationWhenIdeaPluginImportedLast() throws IOException {
+    buildFile << """
+      apply plugin: 'java'
+      apply plugin: 'org.inferred.processors'
+      apply plugin: 'idea'
+
+      idea.processors {
+        outputDir = 'foo'
+        testOutputDir = 'bar'
+      }
+    """
+
+    new File(testProjectDir.newFolder('.idea'), 'compiler.xml') << """
+      <?xml version="1.0" encoding="UTF-8"?>
+      <project version="4">
+        <component name="CompilerConfiguration">
+          <annotationProcessing/>
+        </component>
+      </project>
+    """.trim()
+
+    File testProjectDirRoot = testProjectDir.getRoot()
+    GradleRunner.create()
+            .withProjectDir(testProjectDirRoot)
+            .withArguments("idea", "--stacktrace")
+            .build()
+
+    def xml = testProjectDirRoot.toPath().resolve(".idea/compiler.xml").toFile().text.trim()
+
+    def expected = """
+      <project version="4">
+        <component name="CompilerConfiguration">
+          <annotationProcessing>
+            <profile default="true" name="Default" enabled="true">
+              <sourceOutputDir name="foo"/>
+              <sourceTestOutputDir name="bar"/>
+              <outputRelativeToContentRoot value="true"/>
+              <processorPath useClasspath="true"/>
+            </profile>
+          </annotationProcessing>
+        </component>
+      </project>
+    """.stripIndent().trim()
+
+    assertEquals(expected, xml)
   }
 
   private void assertAutoValueInFile(File file) {

